### PR TITLE
1744: carousel panes

### PR DESCRIPTION
**Pull request by xendk**
#### Link to issue

https://platform.dandigbib.org/issues/1744

#### Description

Rewrite ting_search_carousel to be a proper pane with configuration, rather than a single global carousel.

Also fixes up the admin form.

Includes an unrelated fix to suppress some notices (got in the way of debugging), and migration of the existing pane.

Originally based on https://github.com/ding2/ding2/pull/1007

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

### CODE REVIEW

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 9e56a2cae5f550e9003ac52ba3f152c351e20595
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

 * Commit: 8ff75b4d2ee970952ad4cd34de35c4172148c420
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

 * Commit: e40401a502186acf34345c83b25192a90fa29dd2
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

 * Commit: d711fdd2a0f83b21f25b612dbf4d3cbe4c59871d
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 9e56a2cae5f550e9003ac52ba3f152c351e20595
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

 * Commit: 8ff75b4d2ee970952ad4cd34de35c4172148c420
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

 * Commit: e40401a502186acf34345c83b25192a90fa29dd2
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

 * Commit: d711fdd2a0f83b21f25b612dbf4d3cbe4c59871d
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 398d63c0e6636173568701b79c655a3df36d61e8
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 398d63c0e6636173568701b79c655a3df36d61e8
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

File: modules/ting_search_carousel/plugins/content_types/carousel.inc [hunk 1]
```diff
@@ -31,7 +36,7 @@ function ting_search_carousel_carousel_content_type_render($subtype, $conf, $pan
       $carousels[] = array(
         '#type' => 'ding_carousel',
         '#title' => $search['title'],
-        '#path' => 'ting_search_carousel/results/ajax/' . $index,
+        '#path' => 'ting_search_carousel/results/ajax/' . urlencode($search['query']),
```
Comment (cableman) [inline modules/ting_search_carousel/plugins/content_types/carousel.inc]: Will this url encoding of the search query not give problems with max length in urls as the editors makes crazy long search queries (knowns from e.g eReolen that it already breaks).

Comment (cableman) [reply modules/ting_search_carousel/plugins/content_types/carousel.inc]: Would it be possible to create an hash in cache or something that could be used to id the search query for a given carousel?

Comment (cableman) [reply modules/ting_search_carousel/plugins/content_types/carousel.inc]: I see that the query input field is set to an max of 2048, but that should be minus the domain name and "'ting_search_carousel/results/ajax/'" to ensure that this don't happen.

File: modules/ting_search_carousel/plugins/content_types/carousel.inc [hunk 2]
```diff
@@ -58,8 +63,253 @@ function ting_search_carousel_carousel_content_type_admin_title($subtype, $conf,
...
+function ting_search_carousel_carousel_content_type_theme(&$theme, $plugin) {
+  $theme['ting_search_carousel_admin_form'] = array(
+    'render element' => 'form',
+  );
+}
+
+/**
+ * Theme the queries form in a table.
+ *
+ * No tabledrag, as it doesn't seem to like CTools popup window.
```
Comment (cableman) [inline modules/ting_search_carousel/plugins/content_types/carousel.inc]: But is table-drag not add just before the return statement in this function?

File: modules/ting_search_carousel/ting_search_carousel.install [hunk 3]
```diff
@@ -76,20 +37,36 @@ function ting_search_carousel_update_7003() {
 }
 
 /**
- * Ensure router path for carousel shortcut menu link is correct.
+ * Migrate front page carousel.
  */
-function ting_search_carousel_update_7004() {
```
Comment (cableman) [inline modules/ting_search_carousel/ting_search_carousel.install]: I don't think that we should remove old update functions. If not valid, leave them empty with a short description about why it's empty.

File: modules/ting_search_carousel/plugins/content_types/carousel.inc (hunk 1 above)
Comment (kasperg) [reply modules/ting_search_carousel/plugins/content_types/carousel.inc]: I am for the second option "validation on the admin form that errors out if the ajax path (after URL encoding) exceeds 2048". This ensures that we preserve HTTP semantics (idempotent requests should use GET) which is also good for caching, which is important in these scenarios.

Editors making crazy long textual queries are in my opinion expressing a need which is currently not handled very well by the system. If we want to address that need I think we should consider what other options there might be.

**Pull request merged**